### PR TITLE
Use internal config types in scheduling plugin args

### PR DESCRIPTION
**Pull request by alculquicondor**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Adds decoding of nested `pluginConfig.args` so that we can use internal types in the scheduler.
This way, we are compliant with API best practices and, in the future, we will be able to support args using different versions.

**Which issue(s) this PR fixes**:

Part of #87784

**Special notes for your reviewer**:

API changes are in first commit.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types_pluginargs.go [hunk 1]
```diff
@@ -108,6 +110,13 @@ type UtilizationShapePoint struct {
 	Score int32
 }
 
+// UnmarshalJSON provides case insensitive unmarshalling for the type.
+// TODO remove when copying to v1beta1.
+func (t *UtilizationShapePoint) UnmarshalJSON(data []byte) error {
+	type internal *UtilizationShapePoint
+	return gojson.Unmarshal(data, internal(t))
+}
```
Comment (liggitt) [inline staging/src/k8s.io/kube-scheduler/config/v1alpha2/types_pluginargs.go:118]: is a test needed to demonstrate this works?

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types_pluginargs.go [hunk 2]
```diff
@@ -116,6 +125,13 @@ type ResourceSpec struct {
 	Weight int64
 }
 
+// UnmarshalJSON provides case insensitive unmarshalling for the type.
+// TODO remove when copying to v1beta1.
+func (t *ResourceSpec) UnmarshalJSON(data []byte) error {
+	type internal *ResourceSpec
+	return gojson.Unmarshal(data, internal(t))
+}
```
Comment (liggitt) [inline staging/src/k8s.io/kube-scheduler/config/v1alpha2/types_pluginargs.go:133]: is a test needed to demonstrate this works?

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go [hunk 3]
```diff
@@ -202,3 +218,20 @@ type PluginConfig struct {
...
 	Args runtime.RawExtension `json:"args,omitempty"`
 }
+
+func (c *PluginConfig) decodeNestedObjects(d runtime.Decoder) error {
+	gvk := SchemeGroupVersion.WithKind(c.Name + "Args")
+	obj, parsedGvk, err := d.Decode(c.Args.Raw, &gvk, nil)
+	if err != nil {
+		if runtime.IsNotRegisteredError(err) {
+			// Must be args for out-of-tree plugin.
+			return nil
```
Comment (liggitt) [inline staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go:228]: This doesn't distinguish between a NotRegistered error caused because no kind is registered for `gvk` (which we want to tolerate) and an error caused because their args have a kind/apiVersion specified that could not be decoded (which we don't want to tolerate).

It's ugly, but we could do a dry-run decode to our desired gvk to see if the type exists:

```go
gvk := SchemeGroupVersion.WithKind(c.Name + "Args")
if _, _, err := d.Decode([]byte(`{}`), &gvk, nil); err != nil && runtime.IsNotRegisteredError(err) {
  // no typed kind registered for this plugin's args
  return nil
}
```

then parse normally and treat any error as an actual error:
```go
obj, parsedGvk, err := d.Decode(c.Args.Raw, &gvk, nil)
if err != nil {
  return fmt.Errorf("decoding args for plugin %s: %w", s.Name, err)
}
```

and include more details if the kind doesn't match our expectation:
```go
if parsedGvk.GroupKind() != gvk.GroupKind() {
	return fmt.Errorf("args for plugin %s were not of type %s: %s", c.Name, gvk.GroupKind(), parsedGvk.GroupKind())
}
```

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go [hunk 4]
```diff
@@ -99,6 +101,20 @@ type KubeSchedulerConfiguration struct {
...
 }
 
+// DecodeNestedObjects decodes plugin args for known types.
+func (c *KubeSchedulerConfiguration) DecodeNestedObjects(d runtime.Decoder) error {
+	for i := range c.Profiles {
+		prof := &c.Profiles[i]
+		for j := range prof.PluginConfig {
+			err := prof.PluginConfig[j].decodeNestedObjects(d)
+			if err != nil {
+				return fmt.Errorf("deconding .profiles[%d]: %v", i, err)
```
Comment (liggitt) [inline staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go:111]:
```suggestion
				return fmt.Errorf("decoding .profiles[%d]: %w", i, err)
```

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go [hunk 5]
```diff
@@ -99,6 +101,20 @@ type KubeSchedulerConfiguration struct {
 	Extenders []v1.Extender `json:"extenders"`
 }
 
+// DecodeNestedObjects decodes plugin args for known types.
+func (c *KubeSchedulerConfiguration) DecodeNestedObjects(d runtime.Decoder) error {
```
Comment (liggitt) [inline staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go:105]: do we need a symmetric EncodeNestedObjects so we can round-trip a scheduler config from v1alpha2 -> internal -> v1alpha2 and the args convert and get marshaled correctly?

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go (hunk 3 above)
Comment (liggitt) [reply staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go:228]: make sure we have tests exercising the following:
1. no args (no error)
2. args without kind/apiVersion for a known plugin with a type registered (no error, get internal type)
3. args without kind/apiVersion for a custom plugin that does not have an args type registered (no error, get runtime.Unknown internally)
4. args with kind/apiVersion that match the expected GVK (no error, get internal type)
5. args with known kind/apiVersion that do not match the expected GVK (should get a groupkind mismatch error)
6. args with unknown kind/apiVersion (should get a NotRegistered error)

For the non-error cases, I would also then round-trip the resulting internal config back to v1alpha2 and encode it and make sure the arg types get converted/encoded correctly.

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types_pluginargs.go (hunk 1 above)
Comment (alculquicondor) [reply staging/src/k8s.io/kube-scheduler/config/v1alpha2/types_pluginargs.go:118]: They are in scheme_test.go

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go [hunk 6]
```diff
@@ -202,3 +218,20 @@ type PluginConfig struct {
 	// Args defines the arguments passed to the plugins at the time of initialization. Args can have arbitrary structure.
 	Args runtime.RawExtension `json:"args,omitempty"`
 }
+
+func (c *PluginConfig) decodeNestedObjects(d runtime.Decoder) error {
+	gvk := SchemeGroupVersion.WithKind(c.Name + "Args")
+	obj, parsedGvk, err := d.Decode(c.Args.Raw, &gvk, nil)
```
Comment (Huang-Wei) [inline staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go:224]: Can `c.Args.Raw` be empty slice? If so, `err` would be a plain one with message "empty data".

**Review by Huang-Wei — COMMENTED** [review]
some comments.

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go [hunk 7]
```diff
@@ -139,7 +139,7 @@ func (c *codec) Decode(data []byte, defaultGVK *schema.GroupVersionKind, into ru
 	}
 
 	if d, ok := obj.(runtime.NestedObjectDecoder); ok {
-		if err := d.DecodeNestedObjects(runtime.WithoutVersionDecoder{c.decoder}); err != nil {
+		if err := d.DecodeNestedObjects(runtime.WithoutVersionDecoder{c}); err != nil {
```
Comment (Huang-Wei) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go:142]: Does it make any difference?

Comment (alculquicondor) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go:142]: Yes. Only the codec knows about versions, so it can convert from v1alpha2 to internal.

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go (hunk 5 above)
Comment (alculquicondor) [reply staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go:105]: We don't need conversion from internal to v1alpha2. But I added test to encode from v1alpha2

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go (hunk 4 above)
Comment (alculquicondor) [reply staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go]: Done

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go (hunk 3 above)
Comment (alculquicondor) [reply staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go]: Done

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types_pluginargs.go (hunk 2 above)
Comment (alculquicondor) [reply staging/src/k8s.io/kube-scheduler/config/v1alpha2/types_pluginargs.go:133]: They are in scheme_test.go

**Review by alculquicondor — COMMENTED** [review]
Fixes from review in 2nd commit

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go (hunk 5 above)
Comment (liggitt) [reply staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go:105]: > We don't need conversion from internal to v1alpha2

Are you sure? Does the following work:
1. decode config file with args to internal KubeSchedulerConfiguration (args end up as internal Args objects)
2. encode that internal KubeSchedulerConfiguration object to v1alpha2 (args should get converted back to v1alpha2 and serialized correctly)

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go [hunk 8]
```diff
@@ -202,3 +218,21 @@ type PluginConfig struct {
 	// Args defines the arguments passed to the plugins at the time of initialization. Args can have arbitrary structure.
 	Args runtime.RawExtension `json:"args,omitempty"`
 }
+
+func (c *PluginConfig) decodeNestedObjects(d runtime.Decoder) error {
+	gvk := SchemeGroupVersion.WithKind(c.Name + "Args")
+	// dry-run to detect and skip out-of-tree plugin args.
+	if _, _, err := d.Decode(nil, &gvk, nil); runtime.IsNotRegisteredError(err) {
```
Comment (liggitt) [inline staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go:225]: decoding nil returns success?

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go (hunk 5 above)
Comment (alculquicondor) [reply staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go:105]: With the code as-is, that doesn't work (the nested types are directly encoded from internal, thus get a starting Uppercase). We would need to implement the interface.

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go (hunk 8 above)
Comment (alculquicondor) [reply staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go:225]: With the provided gvk, yes :)

File: staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go (hunk 5 above)
Comment (alculquicondor) [reply staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go:106]: I realized that, with the current implementation, v1alpha2 args don't get gvk, which might be dangerous.
But now there is an issue, because the nested objects get enconded into YAML before
https://github.com/kubernetes/kubernetes/blob/470dfbfc4848cee4897a1b176d20611668820492/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go#L305-L314

**Comment by alculquicondor** [comment]
Again, API changes in the n-1 commits

**Comment by alculquicondor** [comment]
Reverted changes in `versioning.go` and doing semi-manual conversion (using a scheme) in `conversion.go`

File: pkg/scheduler/apis/config/v1alpha2/conversion.go [hunk 9]
```diff
@@ -17,20 +17,93 @@ limitations under the License.
...
+			out.Profiles[i].PluginConfig[j].Args = internalArgs
+		}
+	}
 	return nil
 }
 
 func Convert_config_KubeSchedulerConfiguration_To_v1alpha2_KubeSchedulerConfiguration(in *config.KubeSchedulerConfiguration, out *v1alpha2.KubeSchedulerConfiguration, s conversion.Scope) error {
-	return autoConvert_config_KubeSchedulerConfiguration_To_v1alpha2_KubeSchedulerConfiguration(in, out, s)
+	if err := autoConvert_config_KubeSchedulerConfiguration_To_v1alpha2_KubeSchedulerConfiguration(in, out, s); err != nil {
+		return nil
```
Comment (liggitt) [inline pkg/scheduler/apis/config/v1alpha2/conversion.go:83]:
```suggestion
		return err
```

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go [hunk 10]
```diff
@@ -235,7 +235,7 @@ func (c *codec) doEncode(obj runtime.Object, w io.Writer) error {
 	}
 
 	if e, ok := out.(runtime.NestedObjectEncoder); ok {
-		if err := e.EncodeNestedObjects(runtime.WithVersionEncoder{Version: c.encodeVersion, Encoder: c.encoder, ObjectTyper: c.typer}); err != nil {
+		if err := e.EncodeNestedObjects(runtime.WithVersionEncoder{Version: c.encodeVersion, Encoder: c, ObjectTyper: c.typer}); err != nil {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go:238]: revert this change?

Comment (alculquicondor) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go]: yeah, missed it.

**Comment by alculquicondor** [comment]
@Huang-Wei anything to add?
